File: java/Arrays & Maths/N-3 Repeat Number.java
public class Solution {
    public int repeatedNumber(int[] A) {
        int n = A.length;
        int count1 = 0, count2 = 0;
        int first =  Integer.MIN_VALUE;;
        int second = Integer.MAX_VALUE;
     
        for (int i = 0; i < n; i++) {
            if (first == A[i])
                count1++;
            else if (second == A[i])
                count2++;
            else if (count1 == 0) {
                count1++;
                first = A[i];
            }
            else if (count2 == 0) {
                count2++;
                second = A[i];
            }
            else {
                count1--;
                count2--;
            }
        }
     
        count1 = 0;
        count2 = 0;
     
        for (int i = 0; i < n; i++) {
            if (A[i] == first)
                count1++;
            else if (A[i] == second)
                count2++;
        }
     
        if (count1 > n / 3)
            return first;
     
        if (count2 > n / 3)
            return second;
     
        return -1;
    }
}

File: python/Arrays & Maths/N-3 Repeat Number.py
class Solution:
    # @param A : tuple of integers
    # @return an integer
    def repeatedNumber(self, A):
        n = len(A)
        k = 3
        count = [0]*(k-1)
        cand = [-1]*(k-1)
        for i in range(n):
            j = 0
            # A[i] is checked if in cands
            while j < k-1:
                if cand[j] == A[i]:
                    count[j] += 1
                    break
                j += 1
            if j == k - 1:
                # A[i] is not in cands
                l = 0
                while l < k - 1:
                    if count[l] == 0:
                        cand[l] = A[i]
                        count[l] = 1
                        break
                    l += 1
                if l == k - 1:
                    # No candidate is empty. We need to ignore it and reduce counts by 1
                    for l in range(k-1):
                        count[l] -= 1
        
        # Check actual counts of candidates
        for i in range(k-1):
            temp = cand[i]
            tempCnt = 0
            for j in range(n):
                if A[j] == temp:
                    tempCnt += 1
            if tempCnt > n/k:
                return temp
        return -1